### PR TITLE
Fix KeyVault handling and move Secret types into Azure.Core.Expression.DataFactory

### PR DESCRIPTION
**Pull request by JoshLove-msft**
Extracts the following types from Azure.ResourceManager.DataFactory:
- DataFactoryKeyVaultSecretReference
- DataFactorySecretString
- DataFactorySecretBaseDefinition
- DataFactoryLinkedServiceReference
- DataFactoryLinkedServiceReferenceType

Ideally we would generate these types in the Azure.Core.Expression.DataFactory library, but this can be done as a follow-up.

### CODE REVIEW

File: sdk/core/Azure.Core.Expressions.DataFactory/api/Azure.Core.Expressions.DataFactory.netstandard2.0.cs [hunk 1]
```diff
@@ -25,30 +25,53 @@ public sealed partial class DataFactoryElement<T>
...
+        public static Azure.Core.Expressions.DataFactory.DataFactoryElement<System.String?> FromSecretString(Azure.Core.Expressions.DataFactory.DataFactorySecretString secretString) { throw null; }
         [System.ComponentModel.EditorBrowsableAttribute(System.ComponentModel.EditorBrowsableState.Never)]
         public override int GetHashCode() { throw null; }
         public static implicit operator Azure.Core.Expressions.DataFactory.DataFactoryElement<T> (T literal) { throw null; }
         public override string? ToString() { throw null; }
     }
-    public partial class DataFactoryMaskedString
+    public partial class DataFactoryKeyVaultSecretReference : Azure.Core.Expressions.DataFactory.DataFactorySecretBaseDefinition
     {
-        public DataFactoryMaskedString(bool value) { }
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core.Expressions.DataFactory/api/Azure.Core.Expressions.DataFactory.netstandard2.0.cs:38]: Removing all of the overloads for now - we can always add these in the future.

File: sdk/core/Azure.Core.Expressions.DataFactory/src/DataFactorySecretBaseDefinition.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace Azure.Core.Expressions.DataFactory
+{
+    /// <summary>
+    /// The base definition of a secret type.
+    /// Please note <see cref="DataFactorySecretBaseDefinition"/> is the base class. According to the scenario, a derived class of the base class might need to be assigned here, or this property needs to be casted to one of the possible derived classes.
+    /// The available derived classes include <see cref="DataFactorySecretString"/> and <see cref="DataFactoryKeyVaultSecretReference"/>.
+    /// </summary>
+    [TypeReferenceType]
+    public abstract partial class DataFactorySecretBaseDefinition
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core.Expressions.DataFactory/src/DataFactorySecretBaseDefinition.cs:12]: This type was extracted from the Azure.ResourceManager.DataFactory library. The plan is to have code gen use the type from here by leveraging the TypeReferenceType attribute.

File: sdk/core/Azure.Core.Expressions.DataFactory/src/DataFactoryKeyVaultSecretReference.Serialization.cs [hunk 3]
```diff
@@ -0,0 +1,83 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace Azure.Core.Expressions.DataFactory
+{
+    [JsonConverter(typeof(DataFactoryKeyVaultSecretReferenceConverter))]
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core.Expressions.DataFactory/src/DataFactoryKeyVaultSecretReference.Serialization.cs:10]: JsonConverter was added to allow for public serialization from the Azure.ResouceManager.DataFactory library.

**Review by ArthurMa1978 — APPROVED** [review]

**Review by annelo-msft — COMMENTED** [review]
Would you be willing to add some additional context?  What motivates the change?  Is the library in preview so API changes aren't a concern?  Thanks!

**Comment by JoshLove-msft** [comment]
> Would you be willing to add some additional context? What motivates the change? Is the library in preview so API changes aren't a concern? Thanks!

Sure, this fixes how serialization works when using KeyVault references with the DataFactory service. It is a preview library so API changes are allowed.

**Review by m-nash — APPROVED** [review]

**Pull request merged**